Libraries/Opc.Ua.Server/Subscription/Subscription.cs

```diff
@@ -1965,19 +1965,98 @@ public void ValidateConditionRefresh(OperationContext context)
         /// </summary>
         public void ConditionRefresh()
         {
-            ServerSystemContext systemContext = m_server.DefaultSystemContext.Copy(m_session);
             List<IEventMonitoredItem> monitoredItems = new List<IEventMonitoredItem>();
 
+            lock (m_lock)
+            {
+                // build list of items to refresh.
+                foreach (LinkedListNode<IMonitoredItem> monitoredItem in m_monitoredItems.Values)
+                {
+                    MonitoredItem eventMonitoredItem = monitoredItem.Value as MonitoredItem;
+
+                    if (eventMonitoredItem != null && eventMonitoredItem.EventFilter != null)
+                    {
+                        // add to list that gets reported to the NodeManagers.
+                        monitoredItems.Add(eventMonitoredItem);
+                    }
+                }
+
+                // nothing to do if no event subscriptions.
+                if (monitoredItems.Count == 0)
+                {
+                    return;
+                }
+            }
+
+            ConditionRefresh(monitoredItems, 0);
+        }
+
+
+        /// <summary>
+        /// Refreshes the conditions.
+        /// </summary>
+        public uint ConditionRefresh2(uint monitoredItemId)
+        {
+            uint statusCode = StatusCodes.Good;
+
+            List<IEventMonitoredItem> monitoredItems = new List<IEventMonitoredItem>();
+
+            lock (m_lock)
+            {
+                // build list of items to refresh.
+                if ( m_monitoredItems.ContainsKey(monitoredItemId) )
+                {
+                    LinkedListNode<IMonitoredItem> monitoredItem = m_monitoredItems[monitoredItemId];
+
+                    MonitoredItem eventMonitoredItem = monitoredItem.Value as MonitoredItem;
+
+                    if (eventMonitoredItem != null && eventMonitoredItem.EventFilter != null)
+                    {
+                        // add to list that gets reported to the NodeManagers.
+                        monitoredItems.Add(eventMonitoredItem);
+                    }
+                }
+                else
+                {
+                    return StatusCodes.BadMonitoredItemIdInvalid;
+                }
+
+                // nothing to do if no event subscriptions.
+                if (monitoredItems.Count == 0)
+                {
+                    return statusCode;
+                }
+            }
+
+            ConditionRefresh(monitoredItems, monitoredItemId);
+
+            return statusCode;
+        }
+
+        /// <summary>
+        /// Refreshes the conditions.  Works for both ConditionRefresh and ConditionRefresh2
+        /// </summary>
+        public void ConditionRefresh(List<IEventMonitoredItem> monitoredItems, uint monitoredItemId )
```

Comment:
This method should be marked as _private_ since it should only be called by `ConditionRefresh()` or `ConditionRefresh2(uint monitoredItemId)`

Reply:
Updated as requested